### ISSUE TITLE
[$100] New post notification doesn't update the interface to the appropriate post

### ISSUE THREAD

**Issue by vic-topcoder** [appirio-tech/connect-app#2177]
### Expected behavior
When I click on a new post notification, the system would (re)load the page to show the post, and scroll to the post into view.

### Actual behavior
The interface doesn't load the appropriate post.

Aha! Link: https://topcoder.aha.io/features/TCCONNECT-465

**Comment by Gauravseta** [comment]
@vic-topcoder can i take this? or is it not for open for pickup hence not a part of bug bash?

**Comment by gondzo** [comment]
It's not part of the bug bash @Gauravseta

**Comment by gondzo** [comment]
This can be implemented after #2306 is merged

**Comment by maxceem** [comment]
@vikasrohit for now it doesn't work as expected in this issue.

Here is a demo where `customer` made one more comment but `manager` doesn't see it after clicking on the notification https://monosnap.com/file/PZzZ0p1vOv6QWQwUNxe194xGQGMof9

Note, before the fix for the issue https://github.com/appirio-tech/connect-app/issues/2465 I think this could work automatically but not anymore.

Before fix for https://github.com/appirio-tech/connect-app/issues/2465 we reloaded all the posts on the dashboard tab every time we navigate to it from any other place. But for https://github.com/appirio-tech/connect-app/issues/2465 I have to disable it, see explanation for this change here https://github.com/appirio-tech/connect-app/issues/2465#issuecomment-423019557.

So it looks like to keep smooth behaviour for transactions between tabs when using sidebar and still update posts when we click notifications, we have to add some additional callback for clicks on notifications which will force posts reloading. What do you think? Is there any better solution?

**Comment by vikasrohit** [comment]
@maxceem I guess you linked wrong demo video in the comment. However, I do understand the issue here. Can we reload the topic in the FeedContainer when we see the `#` in the URL? I mean if the URL has `#comment-2345`, can just refresh the topic of the comment? I think doing that in only `componentWillMount` would work as ideally if the user is already on the same page where the notification leads to, that page should automatically ask user to refresh the page because of the availability of the new content.

**Comment by vikasrohit** [comment]
> So I think we have to check if topic/comment is already loaded.

How would we do that? And what would be the cost in terms of time?

IMO, it should not be bad experience to always, without checking if it exists or not, load the specific topic and its posts only when the container is mounted and the URL contains the anchored link.
And as I said, if user is already on the target page, componentWillMount will not be fired and it won't reload but our other system of alerting user on new content availability should either cause the auto refresh or show a button to refresh the content.

**Comment by maxceem** [comment]
> How would we do that? And what would be the cost in terms of time?

We can check that inside `componentWillMount` if topic is already in Redux Store - means it's loaded in the connect app. It has to be very fast as it's searching by id in array.

> IMO, it should not be bad experience to always, without checking if it exists or not, load the specific topic and its posts only when the container is mounted and the URL contains the anchored link.

Maybe we can describe cases separately to make sure we have the same view on this:
1. We click a link on some `dashboard topic` on the sidebar while we are on the `project plan tab`. In this case we have to switch to the `dashboard tab` without page reload to make the transaction smooth. Otherwise, it will look like this https://monosnap.com/file/pv92XeI03cXO9KWaC1Bt3rDuJYjIy9
 2. But when we click on the **NEW** `dashboard topic` in the `notifications dropdown`, this NEW topic is not loaded yet, so when we go the dashboard page, it has to be reloaded to get an updated list of the topics.

**Note** In both cases 1 and 2 we click the link of the same shape `https://connect.topcoder.com/projects/7306#feed-27895` so we cannot undestand from it whether we clicked it on sidebar or notifications dropdown.

@vikasrohit let me know if I'm missing something here.

**Comment by vikasrohit** [comment]
I guess, as we have the topic id in the notification, we can fetch only that specific topic by `GET /v5/topics/{topicId}` call. Do you think that can cause any further troubles?

**Comment by maxceem** [comment]
> I guess, as we have the topic id in the notification, we can fetch only that specific topic by GET /v5/topics/{topicId} call. Do you think that can cause any further troubles?

I think it should work. We only have to make sure, we add this separately loaded topic in the same order as they are usually returned from the server.

So the last my concern is regarding the **b.** When we lick a dashboard topic on the project plan sidebar we will move to the dashboard tab and topics will be remounted. So the topic we've just clicked will be reloaded. It feels that in this case navigation to the topic using the sidebar would be not so smooth. But let me know if reloading topic in such case is ok.

**Comment by vikasrohit** [comment]
>  It feels that in this case navigation to the topic using the sidebar would be not so smooth. But let me know if reloading topic in such case is ok.

I don't think we have any other option here. We have to reload the topic or it posts (depending on the event).

**Comment by maxceem** [comment]
> I don't think we have any other option here. We have to reload the topic or it posts (depending on the event).

@vikasrohit ok, the last though to confirm. When we will navigate by the link `https://connect.topcoder-dev.com/projects/7442#feed-29304` from the email:
- first, all the topics will be loaded
- after a particular topic will be reloaded due to present of `#feed-29304`

Otherwise, I think we can still check if topic is already loaded and don't reload it if it's here.

**Comment by RishiRajSahu** [comment]
>Otherwise, I think we can still check if topic is already loaded and don't reload it if it's here.

**Comment by RishiRajSahu** [comment]
>Otherwise, I think we can still check if topic is already loaded and don't reload it if it's here.

**Comment by vikasrohit** [comment]
@maxceem I am not able to follow now, what exactly was the issue and what we are trying to implement as fix. Could you please summarize the issues and solution in one comment. We would decide from there.
fyi @RishiRajSahu

**Comment by maxceem** [comment]
@vikasrohit 

#### The issue can be reproduced in the next way:
- we open somw project details page (any tab)
- some other user creates a new post either on **Dashboard** or under some **Phase** on the **Project Plan** tab
- we see a notification about the new post in the top right corner notification center
  
   ![image](https://user-images.githubusercontent.com/146016/51094846-5e65e400-17eb-11e9-9e29-ffef847133fe.png)

- we click on the notification, but as the post is not loaded we cannot see it
- Note: all posts now are loaded when we load the project, **Dashboard** or **Project Plan** tab in `ProjectInfoContainer` https://github.com/appirio-tech/connect-app/blob/hotfix/horizontal-layout-for-questions/src/projects/detail/containers/ProjectInfoContainer.js#L58-L70

#### Desired behavior:
- when we click on the notification regarding a new post, the post has to be loaded (if it's not yet) and the browser has to scroll to it

#### Suggested solution:
- listen to the URL changes in some component which is present on all project tabs, I think it would be `ProjectDetail.jsx`.
- if the new URL points to some post (`https://connect.topcoder-dev.com/projects/7682/#feed-30349`) or comment  (`https://connect.topcoder-dev.com/projects/7682/#comment-81200`)
- then we check if this post/comment is already in the Redux store:
   - either **Dashboard** page post/comment:
     
      ![image](https://user-images.githubusercontent.com/146016/51095049-ccf77180-17ec-11e9-8731-591002a872e1.png)

  - or some **Phase** comment:

    ![image](https://user-images.githubusercontent.com/146016/51095071-08923b80-17ed-11e9-8968-c184cc371fb9.png)

- if we can the post in the Redux store - we do nothing, as the navigation to post is already implemented
- if we cannot find the post in the Redux store, we force realoading page. So we do something like:
   ```
    window.location.href = `https://connect.topcoder-dev.com/projects/7682/#feed-30349`
   ```
   This is the simple and robust approach.
- other aproach would be to trigger only Dashborad and Phases posts realoding, same like here https://github.com/appirio-tech/connect-app/blob/hotfix/horizontal-layout-for-questions/src/projects/detail/containers/ProjectInfoContainer.js#L58-L70
   - in this case automatic smooth scrolling to the post which is already implemented most likely wouldn't work, and would have to care about such case separately
   - also, we should check more carefully that if we reload posts like this nothing is broken as there could some code which depends on the posts in the Redux store

**Comment by maxceem** [comment]
Sum up:
This issue can be handled after changes in https://github.com/appirio-tech/connect-app/issues/2575#issuecomment-478834793

#### Main goal
When another user creates a new post on the Dashboard tab or at the Phase we can get a notification about this. But when we will click such notification, we couldn't see the post, because the post is not yet loaded by itself. So when we navigate to some post we have to check if it's already loaded or no. And if not loaded we should trigger post reloading so it would be loaded.

#### Implementation details
- we can use users `pshah_manager` and `pshah_copilot` to test notifications:
   - Create a project with `pshah_copilot`
   - Open the same project as `pshah_manager` and `Join it`.
    - After that you can create posts with one of the users and get notifications as another one.
- At the moment we load posts for Dashboard and Phases in the sidebar container https://github.com/appirio-tech/connect-app/blob/dev/src/projects/detail/containers/ProjectInfoContainer.js#L64-L76
   - we load them only if they were not loaded yet
   - we should update this behaviour  for Dashboard posts:
      - check if URL is to some topic `/projects/{{projectId}}/#feed-{{topicId}}` or comment `/projects/{{projectId}}/#comment-{{postId}}` 
      - check if inside redux store we already have the referred post loaded:

         ![image](https://user-images.githubusercontent.com/146016/55373897-41fb5d00-5539-11e9-9e8e-29ac756ee2eb.png)

     - if topic/post is not loaded yet, we should trigger posts/topic reloading
   - similar update this behaviour for Phases posts:
      - check if URL is to some phase post `/projects/{{projectId}}/plan/#phase-{{phaseId}}-posts-{{postId}}` 
      - check if inside redux store we already have the referred phase and post loaded:

         ![image](https://user-images.githubusercontent.com/146016/55375452-cd2b2180-553e-11e9-8ad2-a643f8981f37.png)

     - if post is not loaded yet, we should trigger phase posts reloading

**Comment by maxceem** [comment]
This issue is not open for pick up.

Please, pull new commits from the

**Comment by nahidshahin** [comment]
@maxceem can I try it, please ?

**Comment by maxceem** [comment]
done

**Comment by maxceem** [comment]
As no PR has been raised during 12 hours and no comment provided this issue is now open for pick up.

**Comment by sagarbaver** [comment]
@maxceem I'd like to take a shot at this one ✋

**Comment by gurmeetb** [comment]
@maxceem I can take it up

**Comment by maxceem** [comment]
@sagarbaver as you the first one who commented on this issue it's on you as per our Bug Bash rules.

To keep it clear I can assign you the issue in the UI. To make it, please, accept the invitation to the repository which you should get on your email.

Thanks.

**Comment by sagarbaver** [comment]
@maxceem Thanks, I've accepted the invite.

**Comment by sagarbaver** [comment]
Hey @maxceem! Just wanted to let you know that I couldn't squeeze out enough time from work to do this under 12 hrs. However, I'm back home and am actively working on this. I could use a couple of hours. 
Thanks.

**Comment by sagarbaver** [comment]
Hi @maxceem! 
> check if URL is to some topic

I believe this is not as straightforward as it appears. The reason being those notification items are react-router `<Link>` elements. A click on them neither triggers the `window.onhashchange` event nor does it trigger an update of the `ProjectInfoContainer.js` component.

I'm thinking of alternative ways to load posts on-demand.

**Comment by maxceem** [comment]
As we use react-router v4, here is how you can listen for URL change https://stackoverflow.com/questions/45373742/detect-route-change-with-react-router/45373907#45373907. Let me know if it covers the case when we change only the hash.

**Comment by sumitdaga** [comment]
@maxceem  @sagarbaver   
can i try this if it's ok?

**Comment by sagarbaver** [comment]
@sumitdaga I'm working on incorporating @maxceem suggestion on leveraging react-router props.

**Comment by sagarbaver** [comment]
@maxceem I wanted to share my progress and gather some thoughts on possible solutions to this issue. Please take a look at https://github.com/appirio-tech/connect-app/pull/2961.

**Comment by RishiRajSahu** [comment]
@maxceem removing it from CF

**Comment by maxceem** [comment]
Sum up:
1. Asynchronously reload posts/topics using actions `loadDashboardFeeds`, `laodProjectMessages` and `loadPhaseFeed`.
   When we reload feed on the dashboard, we should call both `loadDashboardFeeds` (load public topics) and `loadProjectMessages` (load private topics) as we don't know if the new topic if private or public.

   For posts on the dashboard, we can detect if the post is inside private or public topic. If it's private, call `loadProjectMessages` if public, call `loadDashboardFeeds`. (To detect private/public topic we should check `topic.tag`. If `topic.tag === PROJECT_FEED_TYPE_MESSAGES` then it's a private topic, otherwise - public).

   For phases, we use `loadPhaseFeed`.

2. For phases logic would be a little bit more complicated. In case another user creates a new phase and after adds a new post to it. If we load posts for such case it wouldn't work. As we don't have ready actions to load only one phase, to keep is simple we can reload all the phases and their posts in such case (I think it would rare).
    Currently, phases are being loaded here: https://github.com/appirio-tech/connect-app/blob/cf16/src/projects/actions/projectDashboard.js#L47-L69. First we load all the phases, after we load all the posts for phases and after we load users for the posts if necessary.
    We can reuse the same code to reload all the phases in case if the phase with posts which we want to show is not loaded.

   **When implementing this logic, most likely you would need to copy logic from https://github.com/appirio-tech/connect-app/blob/dev/src/projects/actions/projectDashboard.js#L47-L69. Instead, create a new reusable method for this instead and use in both places**

3. We should make sure that we show a loader when we load topics on the dashboard and posts for phases. I think for dashboard it should work automatically during topics loading. But for phases, we should add LoaderIndicator during phase posts loading for one phase.
    Also, we should show a loading indicator when we reload all the phases.

4. Scrolling functionality should keep working. When we click notifications for posts, the page is automatically scrolled to the post. It's done and should keep working. For phases: phase should become opened on the Discussion tab and we should scroll to a particular post. This also already works in case posts are already loaded and it should keep working.

**Comment by mhikeo** [comment]
@maxceem Please assign this to me. Thanks.

**Comment by maxceem** [comment]
done

**Comment by gautam1168** [comment]
@maxceem is this becoming open again?

**Comment by maxceem** [comment]
As there were no PR raised and no comment provided, this issues is now open for pickup. The one who first comments here after it's open can start working on it.

**Comment by sumitdaga** [comment]
Can you assign it to me please ? @maxceem

**Comment by maxceem** [comment]
Another issue is on you @sumitdaga, do you want to give it up?

**Comment by maxceem** [comment]
@sumitdaga since you've released another issue, this issue is on you.

**Comment by sumitdaga** [comment]
@maxceem  
i may be wrong, but it seems when new notifications are loaded and if it has a new post, posts are re-loaded as of now ...can you please check it once?

**Comment by maxceem** [comment]
@sumitdaga you are right, in case if are on the Dashboard tab, didn't scroll yet and didn't start editing any post, all the posts would be reloaded once we get notifications about new post.

But it many other situations it wouldn't happen. Also, unlike here, we don't want to reload posts every time a new notification comes. We want to reload posts only after we clicked some notification in case the according post is not loaded yet.

**Comment by sumitdaga** [comment]
@maxceem  
sorry i couldn't put much time today , but would like to continue to work on it tomorrow..  
let me know if its ok ?

**Comment by maxceem** [comment]
You can continue working on today.

**Comment by maxceem** [comment]
@sumitdaga could you please, let me know if you are working on the issue.

**Comment by sumitdaga** [comment]
@maxceem  yes i am working, if i am not able to finish by today ..i will open it for pickup!

**Comment by sumitdaga** [comment]
@maxceem  please unassign me from this!

**Comment by maxceem** [comment]
This issue is now open for pickup.

**Comment by gurmeetb** [comment]
@maxceem I can try this.

**Comment by maxceem** [comment]
sure

**Comment by gurmeetb** [comment]
@maxceem sorry got pulled into something, can we open this for others? Thanks

**Comment by sumitdaga** [comment]
@maxceem  i would like to work on it again since i have already raised a PR for the other issue i was working on!

**Comment by sumitdaga** [comment]
@maxceem  i am working on it ..meanwhile please give feedback, if any on the approach  at #3035

**Comment by maxceem** [comment]
@sumitdaga I've checked the WIP PR the global idea feels right - tracking changing of the hash. Though we don't have to create a custom way for tracking it we can use `react-router` package and `withRouter` HOC.

Could you please check this [comment](https://github.com/appirio-tech/connect-app/issues/2177#issuecomment-478836122) which describes the general approach here.
Also, check this closed [PR](https://github.com/appirio-tech/connect-app/pull/2961/files) it has partially implemented this task. You may also check comments for that PR. But note, in that PR I asked to skip some functionality, but now it's in scope.

**Comment by sumitdaga** [comment]
@maxceem  
there are some tricky spots with this issue ...  updating projectInfoContainer has as such no issues ..after implementing that part comes scrolling to the relevant feed or post ...  
one way is to move the `scrollTo` functionality to `componentDidUpdate()` lifecycle hook in the https://github.com/appirio-tech/connect-app/blob/dev/src/projects/detail/containers/FeedContainer.js#L223 component  ...this solves the problem but then every time the compoenentUpdates it scrolls to the urlHash ...it happens as of now in the projectPlan tab as well ...   
to avoid this ...we somehow need to either remember that we have already scrolled or some other way , not sure what is the best way ...

**Comment by sumitdaga** [comment]
> this solves the problem but then every time the compoenentUpdates it scrolls to the urlHash   
is this behaviour ok ?

**Comment by sumitdaga** [comment]
@maxceem   i have updated the PR #3035  
let me know if its ok?

**Comment by maxceem** [comment]
> there are some tricky spots with this issue ... updating projectInfoContainer has as such no issues ..after implementing that part comes scrolling to the relevant feed or post ...

I think scrolling should work automatically, doesn't it?

> this solves the problem but then every time the compoenentUpdates it scrolls to the urlHash
> is this behaviour ok ?

This behavior I think would disturb users.

**Comment by sumitdaga** [comment]
> I think scrolling should work automatically, doesn't it?  

when we call the `scrollToHash` method in `componentWillRecieveProps` https://github.com/appirio-tech/connect-app/blob/dev/src/projects/detail/containers/FeedContainer.js#L229  the new feed/post is in the redux store but not yet rendered and it won't scroll ...setTimeout pushes the call a bit futher down the stack during which the feed/post is now rendered and then it scrolls  

also in the project plan tab i couldn't try the `create new phase and then post` flow as i was not able to create a new phase yesterday ...let me know if that works ?

**Comment by maxceem** [comment]
I can create a new phase with user `pshah_manger` here https://connect.topcoder-dev.com/projects/8031/plan. Let me know if it works for you.

**Comment by maxceem** [comment]
Here is a detailed summary regarding the implemented fix for this issue https://github.com/appirio-tech/connect-app/pull/3035#issuecomment-496351914
